x/lockup/keeper/synthetic_lock.go

```diff
@@ -0,0 +1,181 @@
+package keeper
+
+import (
+	"fmt"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/gogo/protobuf/proto"
+	"github.com/osmosis-labs/osmosis/x/lockup/types"
+)
+
+func (k Keeper) setSyntheticLockupObject(ctx sdk.Context, lockID uint64, suffix string, endTime time.Time) error {
+	synthLock := &types.SyntheticLock{
+		LockId:  lockID,
+		Suffix:  suffix,
+		EndTime: endTime,
+	}
+	store := ctx.KVStore(k.storeKey)
+	bz, err := proto.Marshal(synthLock)
+	if err != nil {
+		return err
+	}
+	store.Set(syntheticLockStoreKey(lockID, suffix), bz)
+	if !endTime.Equal(time.Time{}) {
+		store.Set(syntheticLockTimeStoreKey(lockID, suffix, endTime), bz)
+	}
+	return nil
+}
+
+func (k Keeper) deleteSyntheticLockupObject(ctx sdk.Context, lockID uint64, suffix string) {
+	store := ctx.KVStore(k.storeKey)
+	synthLock, _ := k.GetSyntheticLockup(ctx, lockID, suffix)
+	if synthLock != nil && !synthLock.EndTime.Equal(time.Time{}) {
+		store.Delete(syntheticLockTimeStoreKey(lockID, suffix, synthLock.EndTime))
+	}
+	store.Delete(syntheticLockStoreKey(lockID, suffix))
+}
+
+func (k Keeper) GetSyntheticLockup(ctx sdk.Context, lockID uint64, suffix string) (*types.SyntheticLock, error) {
+	synthLock := types.SyntheticLock{}
+	store := ctx.KVStore(k.storeKey)
+	synthLockKey := syntheticLockStoreKey(lockID, suffix)
+	if !store.Has(synthLockKey) {
+		return nil, fmt.Errorf("synthetic lock with ID %d and suffix %s does not exist", lockID, suffix)
+	}
+	bz := store.Get(synthLockKey)
+	err := proto.Unmarshal(bz, &synthLock)
+	return &synthLock, err
+}
+
+func (k Keeper) GetAllSyntheticLockupsByLockup(ctx sdk.Context, lockID uint64) []types.SyntheticLock {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, combineKeys(types.KeyPrefixSyntheticLockup, sdk.Uint64ToBigEndian(lockID)))
+
+	synthLocks := []types.SyntheticLock{}
+	for ; iterator.Valid(); iterator.Next() {
+		synthLock := types.SyntheticLock{}
+		err := proto.Unmarshal(iterator.Value(), &synthLock)
+		if err != nil {
+			panic(err)
+		}
+		synthLocks = append(synthLocks, synthLock)
+	}
+	return synthLocks
+}
+
+func (k Keeper) GetAllSyntheticLockups(ctx sdk.Context) []types.SyntheticLock {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, types.KeyPrefixSyntheticLockup)
+
+	synthLocks := []types.SyntheticLock{}
+	for ; iterator.Valid(); iterator.Next() {
+		synthLock := types.SyntheticLock{}
+		err := proto.Unmarshal(iterator.Value(), &synthLock)
+		if err != nil {
+			panic(err)
+		}
+		synthLocks = append(synthLocks, synthLock)
+	}
+	return synthLocks
+}
+
+// CreateSyntheticLockup create synthetic lockup with lock id and suffix
+func (k Keeper) CreateSyntheticLockup(ctx sdk.Context, lockID uint64, suffix string, unlockDuration time.Duration) error {
+	// Note: synthetic lockup is doing everything same as lockup except coin movement
+	// There is no relationship between unbonding and bonding synthetic lockup, it's managed separately
```

Comment:
Can you explain this more somewhere?

Reply:
On spec document, would be adding few more details for better understanding